feat: Add support for `unittest` output for promtool queries

bradbeam opened the pull request:
This adds support for a new output type `unittest` to make it easier to seed unit test input series values from real data.

I think there's still room for improvement, but figured it'd be worth getting some early feedback. 

```
% go run ./cmd/promtool query instant https://prometheus.demo.do.prometheus.io up     -o unittest                                           
rule_files: []
group_eval_order: []
tests:
- interval: 0s
  input_series:
  - series: up{env="demo", instance="demo.do.prometheus.io:9093", job="alertmanager"}
    values: "1"
  - series: up{instance="demo.do.prometheus.io:8998", job="random"}
    values: "1"
  - series: up{instance="demo.do.prometheus.io:8999", job="random"}
    values: "1"
  - series: up{instance="http://localhost:9100", job="blackbox"}
    values: "1"
  - series: up{env="demo", instance="demo.do.prometheus.io:9100", job="node"}
    values: "1"
  - series: up{instance="demo.do.prometheus.io:3000", job="grafana"}
    values: "1"
  - series: up{instance="demo.do.prometheus.io:8996", job="random"}
    values: "1"
  - series: up{instance="demo.do.prometheus.io:8997", job="random"}
    values: "1"
  - series: up{instance="demo.do.prometheus.io:9090", job="prometheus"}
    values: "1"
  - series: up{instance="localhost:2019", job="caddy"}
    values: "1"
```

```
% go run ./cmd/promtool query range --start=$(( $(date +%s)-60 )) --end=$(date +%s) https://prometheus.demo.do.prometheus.io 'up' -o unittest
rule_files: []
group_eval_order: []
tests:
- interval: 0s
  input_series:
  - series: up{instance="demo.do.prometheus.io:8996", job="random"}
    values: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
      1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
  - series: up{instance="localhost:2019", job="caddy"}
    values: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
      1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
  - series: up{env="demo", instance="demo.do.prometheus.io:9093", job="alertmanager"}
    values: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
      1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
  - series: up{instance="demo.do.prometheus.io:3000", job="grafana"}
    values: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
      1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
  - series: up{instance="demo.do.prometheus.io:8998", job="random"}
    values: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
      1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
  - series: up{instance="demo.do.prometheus.io:8999", job="random"}
    values: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
      1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
  - series: up{instance="demo.do.prometheus.io:9090", job="prometheus"}
    values: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
      1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
  - series: up{instance="http://localhost:9100", job="blackbox"}
    values: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
      1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
  - series: up{env="demo", instance="demo.do.prometheus.io:9100", job="node"}
    values: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
      1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
  - series: up{instance="demo.do.prometheus.io:8997", job="random"}
    values: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
      1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
```

Signed-off-by: Brad Beam <brad.beam@b-rad.info>

bradbeam committed:
feat: Add support for `unittest` output for promtool queries

This adds support for a new output type `unittest` to make it easier to seed
unit test input series values from real data.

Signed-off-by: Brad Beam <brad.beam@b-rad.info>

bradbeam commented on cmd/promtool/main.go:
```diff
@@ -1151,6 +1154,61 @@ func (j *jsonPrinter) printLabelValues(v model.LabelValues) {
+			inputSeries = append(inputSeries, series{Series: metric, Values: strings.Join(value, " ")})
+		}
+
+		yaml.NewEncoder(os.Stdout).Encode(testGroup{Interval: u.Step, InputSeries: inputSeries})
```
Probably worth going up another level with this to `unitTestFile` for easier redirection

roidelapluie commented:
cc @dgl

weyert commented:
Really interesting. I was looking for something like this to generate the unit test mock data. Are you aware of a similar tool that's already readily available that I could use until it is merged?

dgl commented:
Sorry, will get to reviewing this soon.

I suggest if you're interested compile just promtool yourself (as there's
no UI like the rest of Prometheus you can just do: go build ./cmd/promtool)
and try it out -- it really helps on a review not just to have feedback on
the code but whether this fits actual use cases.

On Thu, 2 Feb 2023 at 08:47, Weyert de Boer ***@***.***>
wrote:

> Really interesting. I was looking for something like this to generate the
> unit test mock data. Are you aware of a similar tool that's already readily
> available that I could use until it is merged?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/11680#issuecomment-1412780413>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABKCKOGCK2JW3X3Q4JZ4LWVLKWZANCNFSM6AAAAAASV4X5HY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

beorn7 commented:
Looking at this during the bug scrub.

@dgl @roidelapluie the ball seems to be in your court. Could you commence the review?

In the meantime, @bradbeam , what about the TODOs in this PR? Could you rebase to run the tests again in their current state? One has failed, but it's too long ago too see the actual problem.

bboreham commented:
Hello, this came back around at the bug-scrub.
We commented that this PR is aimed at unit-tests within promtool, but it would also be useful to have a tool which generated the format used by unit tests within the PromQL engine.

@krajorama has kindly offered to pick up the review of this PR.  @bradbeam if you're still around please acknowledge.

krajorama commented on cmd/promtool/main.go:
```diff
@@ -276,6 +276,9 @@ func main() {
 	case "promql":
 		p = &promqlPrinter{}
+	case "unittest":
+		// TODO pick out default value used in query range (L892)
```
Agree on the TODO and also it should actually be done before we can merge this as currently not specifying the `step` leads to zero interval in the output which then defaults to 1m which may or may not be what you want.

I think there's two options:
- move the calculation of the step (and start, end epoch time) out of `QueryRange` so we can reuse the step,
- or just say for MVP that we always write `1m` as the interval, regardless what step was used. That's a bit easier to understand for the user as they don't have to figure out what 0 means.

bradbeam replied:
Does the new approach fall a little closer in line with what you were thinking for the first option?

krajorama commented on cmd/promtool/main.go:
```diff
@@ -1332,6 +1335,57 @@ func (j *jsonPrinter) printLabelValues(v model.LabelValues) {
+				samples[metricName] = []string{}
+			}
+
+			samples[metricName] = append(samples[metricName], strconv.FormatFloat(float64(samplePair.Value), 'f', -1, 64))
```
The query might return samplePair.Histogram != nil, which is a native histogram. I'd just `panic` in the MVP for now to indicate that this is unsupported unless you want to add the support. I'm not actually sure we have a formatter to get you the right output (see https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/#series).

krajorama commented on cmd/promtool/main.go:
```diff
@@ -1332,6 +1335,57 @@ func (j *jsonPrinter) printLabelValues(v model.LabelValues) {
+			}
+
+			for _, samplePair := range stream.Values {
+				samples[metricName] = append(samples[metricName], strconv.FormatFloat(float64(samplePair.Value), 'f', -1, 64))
```
The query might return samplePair.Histogram != nil, which is a native histogram. I'd just `panic` in the MVP for now to indicate that this is unsupported unless you want to add the support. I'm not actually sure we have a formatter to get you the right output (see https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/#series).

krajorama commented on cmd/promtool/main.go:
```diff
@@ -1332,6 +1335,57 @@ func (j *jsonPrinter) printLabelValues(v model.LabelValues) {
+		inputSeries = append(inputSeries, series{Series: metric, Values: strings.Join(value, " ")})
+	}
+
+	// TODO do we want to use `yaml.FutureLineWrap()`
```
is this still true? The help for FutureLineWrap says: ```FutureLineWrap globally disables line wrapping when encoding long strings. This is a temporary and thus deprecated method introduced to faciliate migration towards v3, which offers more control of line lengths on individual encodings, and has a default matching the behavior introduced by this function.```

krajorama commented on cmd/promtool/main.go:
```diff
@@ -145,7 +145,7 @@ func main() {
 
 	queryCmd := app.Command("query", "Run query against a Prometheus server.")
-	queryCmdFmt := queryCmd.Flag("format", "Output format of the query.").Short('o').Default("promql").Enum("promql", "json")
+	queryCmdFmt := queryCmd.Flag("format", "Output format of the query.").Short('o').Default("promql").Enum("promql", "json", "unittest")
```
```suggestion
	queryCmdFmt := queryCmd.Flag("format", "Output format of the query: promql, json, unittest. Unittest output is experimental.").Short('o').Default("promql").Enum("promql", "json", "unittest")
```

krajorama reviewed (changes requested):
I think this would be a nice addition, however there's some gotchas for the end-users, like the `step` calculation and that we don't align results on time. Thus the range query might return 60 samples for one series over an hour period , but return only 5 samples at the end of the interval for a new series. That second series would be written out as if it had samples at the beginning of the interval - you'd need to substitute missing values with `_` https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/#series or at least bail out if you detect the problem.

There might be more traps, so I'd mention in the help text that this is experimental.

bradbeam commented:
> @bradbeam if you're still around please acknowledge.

ack

bradbeam committed:
refactor: Change to using v1.Range for passing start,end,step parameters around

This refactor was introduced to handle parsing query range externally. Since
the printer wanted to account for the various query functions, it made sense
to include them in with this change.

Signed-off-by: Brad Beam <brad.beam@b-rad.info>

bboreham commented:
Closing due to lack of update over many months.
Happy to re-open if someone picks it up again.
